File: de.tototec.tools.cmvn/src/main/java/de/tototec/tools/cmvn/LinePrefixFilterOutputStream.java
package de.tototec.tools.cmvn;

import java.io.FilterOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class LinePrefixFilterOutputStream extends FilterOutputStream {

	private final List<byte[]> filters;
	private List<Integer> curFilter;
	private int curPos = 0;
	private boolean isFirst = true;
	private List<Byte> bytesBuffered = new ArrayList<Byte>();

	public LinePrefixFilterOutputStream(final OutputStream out, final String... filter) {
		super(out);
		this.filters = new ArrayList<byte[]>(filter.length);
		for (final String f : filter) {
			this.filters.add(f.getBytes());
		}
		curFilter = new ArrayList<Integer>(filter.length);
	}

	@Override
	public void write(byte[] b) throws IOException {
		write(b, 0, b.length);
	}

	@Override
	public void write(byte[] b, int off, int len) throws IOException {
		if ((off | len | (b.length - (len + off)) | (off + len)) < 0)
			throw new IndexOutOfBoundsException();

		for (int i = 0; i < len; i++) {
			write(b[off + i]);
		}
	}

	@Override
	public void write(final int b) throws IOException {
		if (isFirst) {
			// Init new line
			for (int i = 0; i < filters.size(); ++i) {
				// add all know filters to a list of actual filters
				curFilter.add(i);
			}
			isFirst = false;
		}

		boolean match = false;

		// End of line
		if (b == '\n') {
			// clear actual filters
			curFilter.clear();
			curPos = 0;
			isFirst = true;

		} else if (!curFilter.isEmpty()) {
			// try to filter current char
			// TODO: use iterator and remove in place
			for (Iterator<Integer> it = curFilter.iterator(); it.hasNext();) {
				byte[] filter = filters.get(it.next());
				if (filter.length > curPos && filter[curPos] == (byte) b) {
					// current byte matches
					match = true;
					if (filter.length == curPos + 1) {
						// exact match (filter-pattern ended here), we can clean
						// buffer
						bytesBuffered.clear();
					} else {
						// buffer the byte
						bytesBuffered.add((byte) b);
					}
				} else {
					// this filter does not match, remove it
					it.remove();
				}
			}

			// if (curFilter.isEmpty() && !bytesBuffered.isEmpty()) {
			// // No matching filter but buffered content, print it out
			// for (final Byte buf : bytesBuffered) {
			// out.write(buf);
			// }
			// bytesBuffered.clear();
			// }

			++curPos;
		}

		if (!match) {
			if (!bytesBuffered.isEmpty()) {
				// buffer not empty, print it out
				for (final Byte buf : bytesBuffered) {
					out.write(buf);
				}
				bytesBuffered.clear();
			}
			out.write(b);
		}
	}

}

File: de.tototec.cmvn/src/main/scala/de/tototec/cmvn/LinePrefixFilterOutputStream.scala
package de.tototec.cmvn

import java.io.FilterOutputStream
import java.io.IOException
import java.io.OutputStream
import java.util.ArrayList
import java.util.Iterator
import java.util.List
//remove if not needed
import scala.collection.JavaConversions._

class LinePrefixFilterOutputStream(out: OutputStream, filter: Array[String]) extends FilterOutputStream(out) {

  def this(out: OutputStream) = this(out, Array())

  private val filters = new ArrayList[Array[Byte]](filter.length)

  private var curFilter: List[java.lang.Integer] = new ArrayList[java.lang.Integer](filter.length)

  private var curPos: Int = 0

  private var isFirst: Boolean = true

  private var bytesBuffered: List[Byte] = new ArrayList[Byte]()

  private var delayedFlush: Boolean = false

  filter foreach { f =>
    this.filters.add(f.getBytes)
  }

  override def write(b: Array[Byte]) {
    write(b, 0, b.length)
  }

  override def write(b: Array[Byte], off: Int, len: Int) {
    if ((off | len | (b.length - (len + off)) | (off + len)) <
      0) throw new IndexOutOfBoundsException()
    delayedFlush = true
    b.slice(off, len).foreach(b => write(b.toInt))
    delayedFlush = false
  }

  override def write(b: Int) {
    if (isFirst) {
      for (i <- 0 until filters.size) {
        curFilter.add(i)
      }
      isFirst = false
    }
    var matched = false
    if (b == '\n') {
      curFilter.clear()
      curPos = 0
      isFirst = true
    } else if (!curFilter.isEmpty) {
      var it = curFilter.iterator()
      while (it.hasNext) {
        var filter = filters.get(it.next())
        if (filter.length > curPos && filter(curPos) == b.toByte) {
          matched = true
          if (filter.length == curPos + 1) {
            bytesBuffered.clear()
          } else {
            bytesBuffered.add(b.toByte)
          }
        } else {
          it.remove()
        }
      }
      curPos
    }
    if (!matched) {
      if (!bytesBuffered.isEmpty) {
        bytesBuffered foreach { buf =>
          out.write(buf)
        }
        bytesBuffered.clear()
      }
      out.write(b)
    }
    if (!delayedFlush) {
      out.flush()
    }
  }

  def setDelayedFlush(delayedFlush: Boolean) {
    this.delayedFlush = delayedFlush
  }
}
